src/lib/ecore/efl_promise2.c

```diff
@@ -0,0 +1,277 @@
+        p->future = f;
+     }
+   else
+     {
+        f->event = ecore_event_add(FUTURE_EVENT_ID, f, _dummy_free, NULL);
+        EINA_SAFETY_ON_NULL_RETURN_VAL(f->event, EINA_FALSE);
+        f->pending_value = value;
+        _pending_futures = eina_list_append(_pending_futures, f);
+     }
+   return EINA_TRUE;
+}
+
+Eina_Bool
+efl_promise2_init(void)
+{
+   const char *choice;
+
+   choice = getenv("EINA_MEMPOOL");
+   if ((!choice) || (!choice[0]))
+     choice = "chained_mempool";
+
+   _promise_mp = eina_mempool_add(choice, "Efl_Promise2",
+                                  NULL, sizeof(Efl_Promise2), 16);
+   EINA_SAFETY_ON_NULL_RETURN_VAL(_promise_mp, EINA_FALSE);
+
+   _future_mp = eina_mempool_add(choice, "Efl_Future2",
+                                 NULL, sizeof(Efl_Future2), 16);
+   EINA_SAFETY_ON_NULL_GOTO(_future_mp, err_future);
+
+   FUTURE_EVENT_ID = ecore_event_type_new();
+
+   _future_event_handler = ecore_event_handler_add(FUTURE_EVENT_ID,
+                                                   _efl_future2_handler_cb,
+                                                   NULL);
+   EINA_SAFETY_ON_NULL_GOTO(_future_event_handler, err_handler);
+
+   return EINA_TRUE;
+
+ err_handler:
+   eina_mempool_del(_future_mp);
+   _future_mp = NULL;
+ err_future:
+   eina_mempool_del(_promise_mp);
+   _promise_mp = NULL;
+   return EINA_FALSE;
+}
+
+void
+efl_promise2_shutdown(void)
+{
+   Efl_Future2 *f;
+
+   EINA_LIST_FREE(_pending_futures, f)
+     _efl_future2_dispatch(f, f->pending_value, f->reject);
+
+   ecore_event_handler_del(_future_event_handler);
+   eina_mempool_del(_future_mp);
+   eina_mempool_del(_promise_mp);
+   _promise_mp = NULL;
+   _future_mp = NULL;
+   _future_event_handler = NULL;
+   FUTURE_EVENT_ID = -1;
+}
+
+EAPI Eina_Value *
+efl_promise2_as_value(Efl_Promise2 *p)
+{
+   Eina_Value *v = eina_value_new(EINA_VALUE_TYPE_PROMISE2);
+   EINA_SAFETY_ON_NULL_RETURN_VAL(v, NULL);
+   eina_value_pset(v, p);
+   return v;
+}
+
+EAPI Eina_Value *
+efl_future2_as_value(Efl_Future2 *f)
+{
+   Eina_Value *v = eina_value_new(EINA_VALUE_TYPE_FUTURE2);
+   EINA_SAFETY_ON_NULL_RETURN_VAL(v, NULL);
+   eina_value_pset(v, f);
+   return v;
+}
+
+EAPI Efl_Promise2 *
+efl_promise2_new(Eina_Free_Cb cancel_cb, const void *data)
+{
+   Efl_Promise2 *p = eina_mempool_calloc(_promise_mp, sizeof(Efl_Promise2));
+   EINA_SAFETY_ON_NULL_RETURN_VAL(p, NULL);
+   p->cancel = cancel_cb;
+   p->data = data;
+   return p;
+}
+
+EAPI void
+efl_promise2_cancel(Efl_Promise2 *p)
+{
+   if (p->cancel)
+     p->cancel((void *)p->data);
+   if (p->future)
+     {
+        if (p->future->event)
+          ecore_event_del(p->future->event);
+        p->future->promise = NULL;
+     }
+   eina_mempool_free(_promise_mp, p);
+}
+
+EAPI Eina_Bool
+efl_promise2_resolve(Efl_Promise2 *p, Eina_Value *value)
+{
+   Eina_Bool r;
+
+   EINA_SAFETY_ON_NULL_RETURN_VAL(p, EINA_FALSE);
+   EINA_SAFETY_ON_NULL_RETURN_VAL(p->future, EINA_FALSE);
```

Comment:
this isn't an error... just free the value.

Also, don't return boolean in this function and always free the value. This will avoid leaks and errors. This includes `EINA_SAFETY_ON_NULL_GOTO(p, free_value)`